feat: add KEDA HTTP Add-on support

Fedosin opened the pull request:
## Add KEDA HTTP Add-on support to the OLM Operator

This PR extends the KEDA OLM Operator to deploy and manage the [KEDA HTTP Add-on](https://github.com/kedacore/http-add-on) (v0.14.0+) as part of the `KedaController` custom resource.

### Key design decisions

- Installation is gated by `spec.httpAddon.enabled` (default: `false`)
- A global `version` field sets the image tag for all components, with per-component overrides via `image.tag`
- `KEDA_HTTP_ADDON_{OPERATOR,INTERCEPTOR,SCALER}_IMAGE` env vars allow downstream OLM image injection, taking precedence over CRD values
- HTTP Add-on CRDs (`InterceptorRoute`, `HTTPScaledObject`) are always installed regardless of the `enabled` flag, to prevent accidental deletion of user-created resources
- Only generic Kubernetes deployment fields are exposed in the CRD; HTTP Add-on-specific tuning is done via environment variables
- HTTP Add-on CRDs are stored in `http-add-on/<version>/manifests/`, separate from the KEDA versioned bundle in `keda/`

### Changes

- Add `HttpAddonSpec`, per-component specs, and `HttpAddonStatus` types to the `KedaController` CRD
- Add `resources/keda-http-addon.yaml` (v0.14.0 upstream manifest) and include it in the Dockerfile
- Add `InterceptorRoute` and `HTTPScaledObject` CRDs to `config/crd/bases/`
- Add `ReplaceContainerEnv`, `ReplaceReplicas`, `ReplaceContainerVolumeMounts`, and generic image/logging transformers
- Add `installHttpAddon` reconciliation with full transform pipeline, including OpenShift-aware seccomp profile handling (pod and container level)
- Add HTTP Add-on cleanup to finalizer with zero-value manifest guard
- Replace hardcoded namespace env vars with `fieldRef` in the manifest
- Add `hack/http-add-on-relprep.sh` release script with generic CRD extraction, kustomization verification, namespace `fieldRef` substitution, and CRD/Namespace stripping
- Update RBAC for `ScaledObject` and `PodDisruptionBudget` resources
- Update sample `KedaController` CR with `httpAddon` section
- Add HTTP Add-on documentation to README

### Usage

Enable the HTTP Add-on by setting `httpAddon.enabled: true` in the KedaController spec:

```yaml
apiVersion: keda.sh/v1alpha1
kind: KedaController
metadata:
  name: keda
  namespace: keda
spec:
  httpAddon:
    enabled: true
    version: "0.14.0"
```

Each component (operator, interceptor, scaler) can be individually configured:

```yaml
spec:
  httpAddon:
    enabled: true
    version: "0.14.0"
    interceptor:
      replicas: 3
      resources:
        requests:
          cpu: 100m
          memory: 128Mi
      env:
        - name: KEDA_HTTP_CONNECT_TIMEOUT
          value: "500ms"
```

The HTTP Add-on is disabled by default and does not affect existing KEDA installations.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)

Fixes #262

linkvt commented on api/keda/v1alpha1/httpaddon_types.go:
```diff
@@ -0,0 +1,438 @@
+
+	// KubeRbacProxy configuration for the operator's kube-rbac-proxy sidecar
+	// +optional
+	KubeRbacProxy HTTPAddonKubeRbacProxySpec `json:"kubeRbacProxy,omitempty"`
```
Should we maybe capitalize `RBAC` to follow the conventions for naming abbreviations, e.g. as with `HTTP`?

linkvt commented on internal/controller/keda/kedacontroller_finalizer.go:
```diff
@@ -28,6 +28,12 @@ func (r *KedaControllerReconciler) finalizeKedaController(logger logr.Logger) er
+	// Delete HTTP Add-on resources if they were installed
+	if err := r.uninstallHTTPAddon(logger); err != nil {
+		logger.Info("error finalized KedaController HTTP Add-on", "error", err)
+		// Don't return error - HTTP Add-on might not have been installed
```
We could check if the error is a not found error (with `errors.IsNotFound`) to not ignore all errors.

linkvt commented on api/keda/v1alpha1/httpaddon_types.go:
```diff
@@ -0,0 +1,438 @@
+	// Set to 0 to disable the operator component
+	// default: 1
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
```
What do you think of adding [validation markers](https://book.kubebuilder.io/reference/markers/crd-validation) to the CRD fields to validate the max, min, enum values, timeout durations, ...?
Not directly only tied to this field but also to port, logging level, ...

linkvt commented on api/keda/v1alpha1/httpaddon_types.go:
```diff
@@ -0,0 +1,438 @@
+}
+
+// HTTPAddonInterceptorAdminSpec defines admin service configuration
+type HTTPAddonInterceptorAdminSpec struct {
```
I see a few issues here and in `HTTPAddonInterceptorProxySpec`
- `Service` seems unused?
- `Port` does only set the env var for the application but doesn't update
	- the deployments ports
	- the "receiving" port in the service
	- the scaler configuration that needs to access the admin port (see `KEDA_HTTP_SCALER_TARGET_ADMIN_PORT`)

Besides that, `Service` should maybe be `ServiceName` to be more clear - or maybe remove it completely, not sure if there is a requirement to rename it?

linkvt commented on api/keda/v1alpha1/httpaddon_types.go:
```diff
@@ -0,0 +1,438 @@
+	Name string `json:"name,omitempty"`
+
+	// Tag is the image tag for kube-rbac-proxy
+	// default: v0.13.0
```
I noticed that the image tag is from 2022. I understand that this tag is still used in the chart, but do we want to deploy this version with the operator?
There are quite a few critical and high findings in the image while `gcr.io/kubebuilder/kube-rbac-proxy` also seems to be deprecated, see https://groups.google.com/g/kubebuilder/c/66NzcUwykpk

There is also this related PR that would us to get rid of kube-rbac-proxy: https://github.com/kedacore/http-add-on/pull/1369

linkvt commented on hack/kedacontroller-with-http-addon.yaml:
```diff
@@ -0,0 +1,81 @@
+# KedaController with HTTP Add-on enabled for testing
```
I think we could simplify this file to only include non default field, as e.g. grpcPort or so is already 9090.
This would make it more clear what we actually need to run the http addon in a minimal deployment.

linkvt commented on resources/keda-http-addon.yaml:
```diff
@@ -0,0 +1,939 @@
+---
```
How was this file generated? Should we add a script/... that allows generating it for updates?

jkyros commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -650,6 +683,280 @@ func (r *KedaControllerReconciler) installMonitoring(ctx context.Context, logger
+	removeSeccomp := util.RunningOnOpenshift(ctx, logger, r.Client) && util.RunningOnClusterWithoutSeccompProfileDefault(logger, r.discoveryClient)
+
+	// --- Operator ---
+	operatorImage := httpAddonResolveImage(addonSpec.Operator.Image, httpAddonDefaultOperatorImage, globalVersion)
```
So how we downstream this, we have some ENV vars we use to override which image we inject: https://github.com/kedacore/keda-olm-operator/blob/331db27583d387930d1f2eecaf7fdef6a4f1eac9/internal/controller/keda/kedacontroller_controller.go#L518

And then that's what we put in the CSV: https://github.com/openshift/custom-metrics-autoscaler-operator/blob/9bd97a48ed8d05c07ae0c8f902cc90c25840318f/hack/cma-generate-csv.sh#L94 

So if we don't expose that interface for the new http add-on-images, we're going to suffer when we pack this downstream. 

I think there are what, 3 of them, so probably: 
```
KEDA_HTTP_ADDON_OPERATOR_IMAGE
KEDA_HTTP_ADDON_INTERCEPTOR_IMAGE
KEDA_HTTP_ADDON_SCALER_IMAGE
```

And those would take precedence over any other image if specified? If we can do that, our existing downstream process will also work for these new "operands"

jkyros commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -650,6 +683,280 @@ func (r *KedaControllerReconciler) installMonitoring(ctx context.Context, logger
+		instance, r.Scheme, logger,
+	)
+	if removeSeccomp {
+		operatorTransforms = append(operatorTransforms, transform.RemoveSeccompProfile(httpAddonContainerOperator, r.Scheme, logger))
```
RemoveSeccompProfile only removes it from the containers not the pods, and the `keda-http-addon.yaml` applies the seccomp profile at the pod level? So this...I think is a noop for http-addon?

 I think we might need to either update `transform.RemoveSecompProfile` to strip them both or align `keda-http-addon.yaml` with the container-level application the other keda manifests do?

(not a blocker, it's gated and I think only affects older kube/OpenShift versions, just mentioning it)

jkyros reviewed (commented):
So generally this looks good. We're going to have to come back for a few things, but I don't think they are blocking. Stuff like: 
- TLS/certificate management for the HTTP add-on components 
- Network Policy (since we added it for everything else) 
- monitoring/ServiceMonitor 

The one real concern I have here is env var consistency + image injection ramifications, and once that's handled I say we merge this so we can ship it with 2.19

Fedosin committed:
feat: add KEDA HTTP Add-on support

Extend the KedaController CRD with an optional `httpAddon` section that
allows deploying the KEDA HTTP Add-on (operator, interceptor, scaler)
as managed components alongside KEDA core.

Key design decisions:
- Installation is gated by `spec.httpAddon.enabled` (default: false)
- A global `version` field sets the image tag for all components,
  with per-component tag overrides via `image.tag`
- KEDA_HTTP_ADDON_{OPERATOR,INTERCEPTOR,SCALER}_IMAGE env vars allow
  downstream OLM image injection, taking precedence over CRD values
- InterceptorRoute and HTTPScaledObject CRDs are always installed
  (via OLM/make install), independent of the enabled flag, to avoid
  destroying user data when the add-on is disabled
- Only generic Kubernetes deployment fields are exposed in the CRD;
  HTTP Add-on-specific tuning is done via env variables
- HTTP Add-on CRDs are stored in http-add-on/<version>/manifests/,
  separate from the KEDA versioned bundle in keda/

Changes:
- Add HttpAddonSpec, per-component specs, and HttpAddonStatus types
- Add resources/keda-http-addon.yaml (v0.14.0 upstream manifest) and
  include it in the Dockerfile
- Add InterceptorRoute and HTTPScaledObject CRDs to config/crd/bases/
- Add ReplaceContainerEnv, ReplaceReplicas, ReplaceContainerVolumeMounts,
  and generic image/logging transformers to transform.go
- Add installHttpAddon reconciliation with full transform pipeline,
  including OpenShift-aware seccomp profile handling (pod and container
  level)
- Add HTTP add-on cleanup to finalizer with zero-value manifest guard
- Replace hardcoded namespace env vars with fieldRef in the manifest
- Add hack/http-add-on-relprep.sh with generic CRD extraction,
  kustomization verification, namespace fieldRef substitution, and
  CRD/Namespace stripping
- Update RBAC for ScaledObject and PodDisruptionBudget resources
- Update sample KedaController CR with httpAddon section
- Add HTTP Add-on documentation to README

Signed-off-by: Mikhail Fedosin <mfedosin@redhat.com>

linkvt commented on config/crd/bases/keda.sh_kedacontrollers.yaml:
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.20.0
+    controller-gen.kubebuilder.io/version: v0.16.5
```
Should we regenerate this file with an up-to-date version of controller-gen?

Fedosin replied:
Done — reinstalled controller-gen from go.mod (v0.20.0) and regenerated all CRDs and RBAC.

linkvt commented on config/rbac/role.yaml:
```diff
@@ -114,3 +138,4 @@ rules:
   - routes
   verbs:
   - '*'
+  - list
```
I think we dont need list as `'*'` exists above.

Fedosin replied:
Good catch — removed the redundant `verbs=list` RBAC marker. The `verbs="*"` in `secret_controller.go` and `configmap_controller.go` already covers it.

linkvt commented on hack/http-add-on-relprep.sh:
```diff
@@ -0,0 +1,141 @@
+
+echo "Updating HTTP Add-on resources for version $ver"
+
+# Download the main manifest from GitHub releases
```
I think we could simplify this quite a lot by using yq, e.g. like `yq 'select(.kind == "CustomResourceDefinition")' manifest.yaml` or to get everything without the namespace and CRDs with `select(.kind != "CustomResourceDefinition" and .kind != "Namespace")` ?

Fedosin replied:
Good suggestion. For now I kept the awk/python approach to avoid adding yq as a dependency, but happy to switch if the project adopts yq as a standard tool.

linkvt commented on hack/http-add-on-relprep.sh:
```diff
@@ -0,0 +1,141 @@
+        r'(- name: ' + env_name + r')\n(\s+)value: \S+',
+        r'\1\n\2valueFrom:\n\2  fieldRef:\n\2    fieldPath: metadata.namespace',
+        result,
+    )
```
We could update the manifests in the addon repo as this is IMO a good default so we would not need this anymore?

Fedosin replied:
Already done! Fixed upstream in https://github.com/kedacore/http-add-on/pull/1625 — the next release will use the downward API natively. Removed the workaround from our script.

linkvt commented on internal/controller/keda/kedacontroller_finalizer.go:
The existing code seems to propagate the errors, should we do the same with `deleteHttpAddon` ?

Fedosin replied:
Done — `deleteHTTPAddon` now returns errors and they are propagated in the finalizer, consistent with the existing pattern.

linkvt commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -109,6 +122,16 @@ func (r *KedaControllerReconciler) SetupWithManager(mgr ctrl.Manager, kedaContro
+
+	httpAddonManifest, err := resources.GetHttpAddonResourcesManifest()
+	if err != nil {
+		logger.Info("HTTP Add-on manifest not found, HTTP Add-on support will be unavailable", "err", err)
```
Shouldn't we be more strict here? I'm not sure if it is a valid case that the HTTP Addon manifests are missing

Fedosin replied:
Agreed — changed to return a hard error (`fmt.Errorf`) instead of a soft info log. The manifest should always be present in the operator image.

linkvt commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -650,6 +683,301 @@ func (r *KedaControllerReconciler) installMonitoring(ctx context.Context, logger
+
+func (r *KedaControllerReconciler) installHttpAddon(ctx context.Context, logger logr.Logger, instance *kedav1alpha1.KedaController, status *kedav1alpha1.KedaControllerStatus) error {
+	if !instance.Spec.HttpAddon.Enabled {
+		logger.Info("HTTP Add-on is disabled, cleaning up resources if present")
```
Wouldn't we log this during every reconciliation? We could check if status.HTTPAddon is not nil which means we would actually uninstall something.

Fedosin replied:
Good point — now we only log and clean up when `status.HTTPAddon != nil`, so it stays silent on every reconciliation when the add-on was never enabled.

linkvt commented on api/keda/v1alpha1/kedacontroller_types.go:
```diff
@@ -167,6 +170,160 @@ type KedaAdmissionWebhooksSpec struct {
+}
+
+// HttpAddonStatus defines the observed state of the HTTP Add-on.
+type HttpAddonStatus struct {
```
The KEDA Status has helper functions like 
```go
func (kcs *KedaControllerStatus) MarkInstallFailed(r string) {
	kcs.Phase = PhaseFailed
	kcs.Reason = r
}
```

We could implement them as well to stay consistent?

Fedosin replied:
Done — added `MarkInstallFailed()` and `MarkInstallSucceeded()` helpers to `HTTPAddonStatus` and use them throughout `installHTTPAddon`.

linkvt commented on api/keda/v1alpha1/kedacontroller_types.go:
```diff
@@ -50,6 +50,9 @@ type KedaControllerSpec struct {
 	ServiceAccount KedaServiceAccountSpec `json:"serviceAccount"`
 
+	// +optional
+	HttpAddon HttpAddonSpec `json:"httpAddon,omitempty"`
```
I think we should use `HTTPAddon...` instead of `HttpAddon...` to follow go naming conventions.

Fedosin replied:
Done — renamed all Go types and fields to `HTTPAddon` (`HTTPAddonSpec`, `HTTPAddonStatus`, `HTTPAddonImageSpec`, etc.). JSON tags stay `httpAddon` per camelCase convention.

linkvt commented on internal/controller/keda/kedacontroller_controller.go:
```diff
@@ -72,6 +73,14 @@ const (
+	httpAddonContainerInterceptor = "interceptor"
+	httpAddonContainerScaler      = "scaler"
+
+	httpAddonDefaultOperatorImage    = "ghcr.io/kedacore/http-add-on-operator"
+	httpAddonDefaultInterceptorImage = "ghcr.io/kedacore/http-add-on-interceptor"
+	httpAddonDefaultScalerImage      = "ghcr.io/kedacore/http-add-on-scaler"
```
I did not find images like this for hte existing code, it's fine for but wanted to make sure we add this intentionally.

Fedosin replied:
Yes, this is intentional. The existing KEDA components get their images injected via env vars (`KEDA_OPERATOR_IMAGE`, etc.) by OLM for disconnected environments. These constants serve as fallback defaults when neither the env var nor the CRD field is set.

linkvt reviewed (approved):
LGTM, the other discussions from John seem resolved and I also checked Copilot comments which were mostly outdated :+1:
